include/RAJA/policy/hip/atomic.hpp

```diff
@@ -55,200 +55,244 @@ using hip_atomicCommon_builtin_types = ::camp::list<
+};
 
-template <typename T,
-          std::enable_if_t<!std::is_arithmetic<T>::value &&
-                           !std::is_enum<T>::value &&
-                           sizeof(T) == sizeof(unsigned short int) &&
-                           sizeof(unsigned short int) == 2, bool> = true>
-RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
-{
-  return RAJA::util::reinterp_A_as_B<unsigned short int, T>(
-    hip_atomicLoad(reinterpret_cast<unsigned short int*>(acc)));
-}
 
+/*!
+ * Type trait for determining if the operator should be implemented
+ * by reinterpreting inputs to types that intrinsics support
+ */
+template <typename T>
+struct hip_useReinterpretLoadStore {
+  static constexpr bool value =
+    !std::is_integral<T>::value &&
+    !std::is_enum<T>::value &&
+    ((sizeof(T) == 1
+#if !defined(UINT8_MAX)
+      && sizeof(unsigned char) == 1
+#endif
+     ) ||
+     (sizeof(T) == 2
+#if !defined(UINT16_MAX)
+      && sizeof(unsigned short) == 2
+#endif
+     ) ||
+     (sizeof(T) == 4
+#if !defined(UINT32_MAX)
+      && sizeof(unsigned int) == 4
+#endif
+     ) ||
+     (sizeof(T) == 8
+#if !defined(UINT64_MAX)
+      && sizeof(unsigned long long) == 8
 #endif
+     ));
 
+  using type =
+    std::conditional_t<sizeof(T) == 1,
+#if defined(UINT8_MAX)
+                       uint8_t,
+#else
+                       unsigned char,
+#endif
+    std::conditional_t<sizeof(T) == 2,
+#if defined(UINT16_MAX)
+                       uint16_t,
+#else
+                       unsigned short,
+#endif
+    std::conditional_t<sizeof(T) == 4,
 #if defined(UINT32_MAX)
-
-template <typename T,
-          std::enable_if_t<!std::is_arithmetic<T>::value &&
-                           !std::is_enum<T>::value &&
-                           sizeof(T) == sizeof(uint32_t), bool> = true>
-RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
-{
-  return RAJA::util::reinterp_A_as_B<uint32_t, T>(
-    hip_atomicLoad(reinterpret_cast<uint32_t*>(acc)));
-}
-
+                       uint32_t,
 #else
+                       unsigned int,
+#endif
+#if defined(UINT64_MAX)
+                       uint64_t>>>;
+#else
+                       unsigned long long>>>;
+#endif
+};
 
-template <typename T,
-          std::enable_if_t<!std::is_arithmetic<T>::value &&
-                           !std::is_enum<T>::value &&
-                           sizeof(T) == sizeof(unsigned int) &&
-                           sizeof(unsigned int) == 4, bool> = true>
-RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
-{
-  return RAJA::util::reinterp_A_as_B<unsigned int, T>(
-    hip_atomicLoad(reinterpret_cast<unsigned int*>(acc)));
-}
+/*!
+ * Alias for determining the integral type of the same size as the given type
+ */
+template <typename T>
+using hip_useReinterpretLoadStore_t = typename hip_useReinterpretLoadStore<T>::type;
 
 #endif
 
-#if defined(UINT64_MAX)
+
+/*!
+ * Atomic load
+ */
+#if defined(__has_builtin) && __has_builtin(__hip_atomic_load)
```

Comment:
This conditional is a bit different than above where store is also checked.

Reply:
That's intentional - this section is just for atomic load. There is a section further down for atomic store. But the type traits are the same for both atomic load and store, so if only one intrinsic is provided, we still want the type traits defined.

Reply:
Whoops, I read the one above as an && instead of an ||.